appcypher-syntax.md

```diff
@@ -0,0 +1,947 @@
+    var op_code = 0b10110001 // Binary literal
+    var interest = 0o56676 // Octal literal
+    ```
+
+## TUPLES
+- Tuple is a special datatype with statically-known length. It allows for cool features like arguments unpacking.
+
+    ```swift
+    var fave_game = ("Uncharted 4", 2016)
+    ```
+
+    One-element tuple. Note the trailing comma.
+
+    ```rust
+    let map = (london,)
+    ```
+
+    Empty tuple.
+
+    ```swift
+    let empty_tuple = ()
+    ```
+
+- Tuple indexing. Tuple index space starts at index 0.
+
+    ```rust
+    let book = ("Harry Potter", "J.K. Rowling", 1995)
+    let title = book.1 // Harry Potter
+    let author = book.2 // J.K. Rowling
+    let year = book.3 // 1995
+    ```
+
+- Tuple destructuring.
+
+    ```rust
+    let person = ("Sam", 50)
+    let (name, age) = person
+    ```
+
+- Tuple unpacking.
+
+    ```swift
+    func add(a, b) {
+        a + b
+    }
+
+    let arguments = (5, 6)
+    add(...arguments)
+    ```
+
+- Tuples have limitations because they are expected to be amenable to static analysis.
+
+    ```swift
+    var tup1 = (1, 2)
+    var tup2 = (5, 6)
+    ```
+
+    Directly or indirectly, a tuple can't be used to extend itself.
+
+    ```go
+    for i in range(x) {
+        tup2 = (...tup2, ...tup1) // Invalid!
+    }
+
+    for i in range(x) {
+        tup3 = (...tup2, 1)
+        tup2 = (tup3, tup1) // Invalid!
+    }
+    ```
+
+- The type of a tuple.
+
+    Since tuples can't change in size, a tuple has a fixed type. For example the tuple `(0xFF, "Hi")` has type `(Int, String)`. `("Hello", 5.0, "Hi")` has type `(String, F64, String)`
+
+## LISTS
+- List is a dynamic container for storing and accessing items.
+
+    ```swift
+    var unordered_list = [7, 3, 8, 5, 4, 0, 9, 1, 2, 6]
+    ```
+
+    Empty list.
+
+    ```swift
+    var empty_list = []
+    ```
+
+    Using List constructor.
+
+    ```swift
+    let students = List::new()
+    ```
+
+- Type annotation for lists.
+
+    ```swift
+    var names: [String] = []
+    ```
+
+- Ratio lists can have heterogenous elements.
+
+    ```swift
+    var stuffs = [500, "Steve", 1.0] // Inferred
+    var stuffs: [F64|String|Int] = [500, "Steve", 1.0] // Explicit typing
+    ```
+
+- List indexing. List index space starts at index 0.
+
+    ```swift
+    stuffs[0] // 500
+    ```
+
+- Slicing a list.
+
+    ```swift
+    contestants[2..7] // 3rd to the 7th index
+    contestants[2..] // 3rd to the last index
+    contestants[..7] // first to the 7th index
+    contestants[..] // first to the last index
+    ```
+
+    A slice can have a step value.
+
+    ```swift
+    let odds = numbers[1..2..30] // The step value is the `2` in the middle.
+    ```
+
+## RANGE
+- Range is an iterable that allows iterating through a range of values.
+
+    ```swift
+    let days = 1..31 // 1 to 30
+    ```
+
+- A range can have a step value.
+
+    ```swift
+    let evens = 2..2..20 // The step value is the `2` in the middle.
+    ```
```

Comment:
This conflicts with the generics _application_ syntax. 

```
func new[T]() -> Self { ... }

let buf = Buffer::new:[Int]()
```

The reason I chose this generics syntax over the typical `Buffer::new<Int>()` syntax, is that it keeps our grammar in the CFG zone. C++ parser is complicated for a reason and it is because it is context-aware which sucks, because it makes parser really hard to write a grammar for and to implement.

Here is why C++ template syntax is context aware. It's easy for the parser to confuse a '<' as a lesser than operator.

```
a < b > ( 1 )
```

This is ambiguous. It can be interpreted as `(a) < (b) > (1)` or as generics _application_ `a<b>(1)` where `b` is a type.

So languages like rust adds a distinction factor, the generics _application_ syntax starts with `::`. So instead of `Buffer::new<Int>()`, you write `Buffer::new::<Int>()`. They call it turbofish, because it looks like a fish. `::<>`

I could have gone with the same turbofish, but I don't find it aesthetically pleasing, so I went for more succinct `Buffer::new:[Int]()`

So that is where the `:` operator is used, and it will ambiguous to use it for range as well. But I guess we could make an exception here. Any colon followed by `[` is a generics _application_